pyproject.toml

```diff
@@ -1,2 +1,94 @@
 [tool.black]
 target-version = ['py310']
+
+# This pyproject.toml is setup so it can be used with or without poetry and also
+# supports editable installs (PEP 660) without breaking IDE and linter inspection.
+
+# DEVELOPER INSTALL INSTRUCTIONS (assuming current directory is at the top of the project) :
+# Poetry:
+#   poetry install --extras dev
+# PIP:
+#   pip install -e .[dev]
+
+[tool.poetry]
+name = "screenpy"
+version = "4.0.2"
+description = "Screenplay pattern base for Python automated UI test suites."
+authors = ["Perry Goy <perry.goy@gmail.com>"]
+maintainers = ["Gabe Langton", "Marcel Wilson"]
+license = "MIT"
+repository = "https://github.com/ScreenPyHQ/screenpy"
+documentation = "https://screenpy-docs.readthedocs.io"
+readme = "README.md"
+classifiers = [
+    "Operating System :: OS Independent",
+    "Programming Language :: Python :: 3 :: Only",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Developers",
+    "Topic :: Software Development :: Quality Assurance",
+    "Topic :: Software Development :: Testing",
+    "Topic :: Software Development :: Testing :: BDD",
+    "Natural Language :: English",
+    "License :: OSI Approved :: MIT License",
+]
+
+# It's possible to add optional dependencies with the poetry CLI tool using:
+#   poetry add --optional some_dev_pkg
+# This will set `optional = true` flag in [tool.poetry.dependencies], as seen below
+# But then you need to remember to manually edit the [tool.poetry.extras] dev section 
+# and declare the package. This allows `pip install .[dev]` to work as expected
+# Poetry users will need to use the `--extras dev` option rather than the `--with dev`
+# so we dont have two different sets of package versions to update.
+
+[tool.poetry.dependencies]
+python = "^3.7"
+PyHamcrest = ">=2.0.0"
+typing-extensions = ">=4.1.1"
+
+# convenience packages for development of screenpy only
+pre-commit = {version = "^2.17.0", optional = true}
+pytest = {version = "^7.1.1", optional = true}
+tox = {version = "^3.26.0", optional = true}
+mypy = {version = "^0.971", optional = true}
+black = {version = "^22.8.0", optional = true}
+isort = {version = "^5.10.1", optional = true}
+flake8 = {version = "^5.0.4", optional = true}
+pylint = {version = "^2.12.2", optional = true}
+coverage = {version = "*", optional = true}
+
+[tool.poetry.extras]
+allure = ["screenpy_adapter_allure"]
+appium = ["screenpy_appium"]
+playwright = ["screenpy_playwright"]
+pyotp = ["screenpy_pyotp"]
+requests = ["screenpy_requests"]
+selenium = ["screenpy_selenium"]
+dev = [
+    "pre-commit",
+    "pytest",
+    "tox",
+    "pylint",
+]
+dev_all = [
+    "pre-commit",
+    "pytest",
+    "tox",
+    "pylint",
+    "mypy",
+    "black",
+    "isort",
+    "flake8",
+    "coverage",
+]
```

Comment:
So... i _think_ you can pin versions in the `extras`. The way this is written, i think the pins above will be ignored if you do `pip install screenpy[dev_all]`, and the latest packages will just be installed. 

I would once again like to voice that i don't think this is the right place for these. If you really want to include the `pre-commit` tools in a convenience file, i would much prefer them to be in `requirements-dev.txt` or something similar. Putting these here sends the wrong signal, in my opinion.

Reply:
Another fabulous feature poetry allows (which is dependent on the extras being configured) is you can uninstall package sets.  

Let's say I installed screenpy with requests `poetry install --extras requests` or `pip install -e .[requests]`  and I no longer need the packages that came with that.  I would only need to run `poetry install` to put me back at the base packages.  There is even an option in poetry to remove any packages that may have been installed via pip that are not part of the pyproject.toml; `poetry install --sync`. 

I get that poetry is not the primary concern, but wanted to point out that having extras configured allows for additional features to work.

Reply:
> There are a couple of problems with using pip to create the requirements.txt file.

Well, i didn't say we should use `pip` to generate the file. We could just write it ourselves. It's just a `.txt` file. Just put the pre-commit packages in there manually. :P

>>    Putting these here sends the wrong signal

> Can you elaborate on this?
> I would ask; is there harm in having a dev packages section configured?

Maybe conventions around the `pyproject` file are still being solidified, but to me, listing `dev` as an extra feels like it's telling folks just installing `screenpy` for funsies that the `dev` package should be available to them. It feels a little like listing "kitchen" on the menu at a restaurant or something, and ordering that would give the customer the pans, knives, oven, etc.

But... like i said, maybe conventions are still being solidified here. I've been treating `pyproject` like an abstraction to the code i want our users to be able to access, so putting the dev tools in there felt like a violation of that abstraction. But maybe like you're saying, it should be the single source of truth for everything about this package, from development to usage (even though `pre-commit` precludes that).

Reply:
> Maybe conventions around the pyproject file are still being solidified

Listing dev as an extra has being done long before the pyproject.toml.  In a few of the examples listed above you can see projects have been putting them in the setup.cfg. 

>  It feels a little like listing "kitchen" on the menu

> I've been treating pyproject like an abstraction to the code i want our users to be able to access

I believe that's a good way to think of it, but I also think it can be more than just a way for _users_ to install their dependencies.  To modify your kitchen analogy a little, I think of it as when you order your meal, you're also asking for the recipe.  (maybe not the best analogy since most restaurant's are not likely to offer "extras" akin to python packages, but I think it suffices)